File: sol/103/103.py
from collections import deque

class Solution:
    def zigzagLevelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        res = []
        if root is None:
            return res
        q = deque([root])
        fromFront = True
        while (s := len(q)) != 0:
            lvl = []
            for _ in range(s):
                if fromFront:
                    node = q.popleft()
                    lvl.append(node.val)
                    if node.left is not None:
                        q.append(node.left)
                    if node.right is not None:
                        q.append(node.right)
                else:
                    node = q.pop()
                    lvl.append(node.val)
                    if node.right is not None:
                        q.appendleft(node.right)
                    if node.left is not None:
                        q.appendleft(node.left)
            fromFront = not fromFront
            res.append(lvl.copy())
        return res

File: sol/103/103.cpp
#include <deque>

struct Solution {
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        vector<vector<int>> res;
        if (root == nullptr){
            return res;
        }
        deque<TreeNode*> q;
        q.push_back(root);
        bool fromFront = true;
        while (!q.empty()){
            vector<int> lvl;
            int s = q.size();
            while (s-- > 0){
                if (fromFront){
                    TreeNode *node = q.front();
                    q.pop_front();
                    lvl.push_back(node->val);
                    if (node->left != nullptr){
                        q.push_back(node->left);
                    }
                    if (node->right != nullptr){
                        q.push_back(node->right);
                    }
                } else {
                    TreeNode *node = q.back();
                    q.pop_back();
                    lvl.push_back(node->val);
                    if (node->right != nullptr){
                        q.push_front(node->right);
                    }
                    if (node->left != nullptr){
                        q.push_front(node->left);
                    }
                }
            }
            res.push_back(lvl);
            fromFront = !fromFront;
        }
        return res;
    }
};
